velox/exec/PrefixSortAlgorithm.h

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+
+#include "velox/common/base/Exceptions.h"
+
+namespace facebook::velox::exec {
+
+struct PrefixSortIterator {
+  PrefixSortIterator(char* prefix, const uint64_t entrySize)
+      : prefix(prefix), entrySize(entrySize) {}
+
+  PrefixSortIterator(const PrefixSortIterator& other)
+      : prefix(other.prefix), entrySize(other.entrySize) {}
+
+  char* operator*() const;
+
+  PrefixSortIterator& operator++();
+
+  PrefixSortIterator& operator--();
+
+  PrefixSortIterator operator++(int);
+
+  PrefixSortIterator operator--(int);
+
+  PrefixSortIterator operator+(const uint64_t i) const;
+
+  PrefixSortIterator operator-(const uint64_t i) const;
+
+  PrefixSortIterator& operator=(const PrefixSortIterator& other);
+
+  uint64_t operator-(const PrefixSortIterator& other) const;
+
+  bool operator<(const PrefixSortIterator& other) const;
+
+  bool operator>(const PrefixSortIterator& other) const;
+
+  bool operator>=(const PrefixSortIterator& other) const;
+
+  bool operator<=(const PrefixSortIterator& other) const;
+
+  bool operator==(const PrefixSortIterator& other) const;
+
+  bool operator!=(const PrefixSortIterator& other) const;
+
+ private:
+  char* prefix;
+  const uint64_t entrySize;
+};
+
+using SwapBufferPtr = std::unique_ptr<char[], std::default_delete<char>>;
+
+inline SwapBufferPtr makeSwapBuffer(uint64_t __n) {
+  return std::unique_ptr<char[], std::default_delete<char>>(new char[__n]());
+}
+
+struct PrefixSortContext {
+  PrefixSortContext(uint64_t entrySize, char* end)
+      : entrySize(entrySize),
+        iterSwapBufPtr(makeSwapBuffer(entrySize)),
+        iterSwapBuf(iterSwapBufPtr.get()),
+        end(end) {}
+  const uint64_t entrySize;
```

Comment:
Should these be private? Perhaps, change PrefixSortContext from struct to class.